{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ac07c8e6-6867-4623-8652-56684e5804b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b9abd385-902e-4d61-81cd-b6e0341464cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from glob import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1112f765-58bc-44a7-adc9-91e5efa55a0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from benchmark_utils import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba29eb1-d17f-444d-82a0-b88f0231d2ec",
   "metadata": {},
   "source": [
    "# No CNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7a1ddb6b-4619-4ec2-b739-004cefcd2c1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "call_base_dir=\"../immunotyper-output/20x\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c166e2ab-9b1c-46b7-9edb-0f67507e6de9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:benchmark_utils:No genotype calls file found for HG002. Pattern: ../immunotyper-output/20x/ighv/HG002.final*IGHV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG005. Pattern: ../immunotyper-output/20x/ighv/HG005.final*IGHV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG01123. Pattern: ../immunotyper-output/20x/ighv/HG01123.final*IGHV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02109. Pattern: ../immunotyper-output/20x/ighv/HG02109.final*IGHV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02486. Pattern: ../immunotyper-output/20x/ighv/HG02486.final*IGHV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02559. Pattern: ../immunotyper-output/20x/ighv/HG02559.final*IGHV_functional_allele_calls.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading annotations for IGHV from ../../digger-functional-annotations-all/\n",
      "Gene type: IGHV\n",
      "Number of samples: 47\n",
      "Number of unique functional alleles: 170\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:benchmark_utils:No genotype calls file found for NA21309. Pattern: ../immunotyper-output/20x/ighv/NA21309.final*IGHV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG002. Pattern: ../immunotyper-output/20x/iglv/HG002.final*IGLV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG005. Pattern: ../immunotyper-output/20x/iglv/HG005.final*IGLV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG01123. Pattern: ../immunotyper-output/20x/iglv/HG01123.final*IGLV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02109. Pattern: ../immunotyper-output/20x/iglv/HG02109.final*IGLV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02486. Pattern: ../immunotyper-output/20x/iglv/HG02486.final*IGLV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02559. Pattern: ../immunotyper-output/20x/iglv/HG02559.final*IGLV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for NA21309. Pattern: ../immunotyper-output/20x/iglv/NA21309.final*IGLV_functional_allele_calls.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading annotations for IGLV from ../../digger-functional-annotations-all/\n",
      "Gene type: IGLV\n",
      "Number of samples: 47\n",
      "Number of unique functional alleles: 69\n",
      "\n",
      "Loading annotations for IGKV from ../../digger-functional-annotations-all/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:benchmark_utils:No genotype calls file found for HG002. Pattern: ../immunotyper-output/20x/igkv/HG002.final*IGKV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG005. Pattern: ../immunotyper-output/20x/igkv/HG005.final*IGKV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG01123. Pattern: ../immunotyper-output/20x/igkv/HG01123.final*IGKV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02109. Pattern: ../immunotyper-output/20x/igkv/HG02109.final*IGKV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02486. Pattern: ../immunotyper-output/20x/igkv/HG02486.final*IGKV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02559. Pattern: ../immunotyper-output/20x/igkv/HG02559.final*IGKV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for NA21309. Pattern: ../immunotyper-output/20x/igkv/NA21309.final*IGKV_functional_allele_calls.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene type: IGKV\n",
      "Number of samples: 47\n",
      "Number of unique functional alleles: 80\n",
      "\n",
      "Loading annotations for TRAV from ../../digger-functional-annotations-all/\n",
      "Gene type: TRAV\n",
      "Number of samples: 47\n",
      "Number of unique functional alleles: 59\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:benchmark_utils:No genotype calls file found for HG002. Pattern: ../immunotyper-output/20x/trav/HG002.final*TRAV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG005. Pattern: ../immunotyper-output/20x/trav/HG005.final*TRAV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG01123. Pattern: ../immunotyper-output/20x/trav/HG01123.final*TRAV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02109. Pattern: ../immunotyper-output/20x/trav/HG02109.final*TRAV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02486. Pattern: ../immunotyper-output/20x/trav/HG02486.final*TRAV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02559. Pattern: ../immunotyper-output/20x/trav/HG02559.final*TRAV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for NA21309. Pattern: ../immunotyper-output/20x/trav/NA21309.final*TRAV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG002. Pattern: ../immunotyper-output/20x/trbv/HG002.final*TRBV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG005. Pattern: ../immunotyper-output/20x/trbv/HG005.final*TRBV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG01123. Pattern: ../immunotyper-output/20x/trbv/HG01123.final*TRBV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02109. Pattern: ../immunotyper-output/20x/trbv/HG02109.final*TRBV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02486. Pattern: ../immunotyper-output/20x/trbv/HG02486.final*TRBV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02559. Pattern: ../immunotyper-output/20x/trbv/HG02559.final*TRBV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for NA21309. Pattern: ../immunotyper-output/20x/trbv/NA21309.final*TRBV_functional_allele_calls.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading annotations for TRBV from ../../digger-functional-annotations-all/\n",
      "Gene type: TRBV\n",
      "Number of samples: 47\n",
      "Number of unique functional alleles: 61\n",
      "\n",
      "Loading annotations for TRGV from ../../digger-functional-annotations-all/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:benchmark_utils:No genotype calls file found for HG002. Pattern: ../immunotyper-output/20x/trgv/HG002.final*TRGV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG005. Pattern: ../immunotyper-output/20x/trgv/HG005.final*TRGV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG01123. Pattern: ../immunotyper-output/20x/trgv/HG01123.final*TRGV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02109. Pattern: ../immunotyper-output/20x/trgv/HG02109.final*TRGV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02486. Pattern: ../immunotyper-output/20x/trgv/HG02486.final*TRGV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG02559. Pattern: ../immunotyper-output/20x/trgv/HG02559.final*TRGV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for NA21309. Pattern: ../immunotyper-output/20x/trgv/NA21309.final*TRGV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG00438.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG005.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG005.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG00621.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG00673.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG00735.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG00741.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG00741.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01106.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01106.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01109.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01123.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01123.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01175.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01175.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01243.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01258.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01258.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01358.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG01928.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02055.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02055.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02080.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02109.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02145.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02148.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02148.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02257.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02486.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02486.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02559.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02559.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02572.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02622.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02630.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02630.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02717.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02723.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02723.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02818.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG02886.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03098.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03098.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03453.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03453.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03486.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03492.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03516.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03540.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/HG03579.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/NA18906.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/NA19240.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/NA20129.paternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:Skipping missing gene file: ../../digger-functional-annotations-all/NA21309.maternal.f1_assembly_v2_genbank/TRD_functional_alleles.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for HG002. Pattern: ../immunotyper-output/20x/trdv/HG002.final*TRDV_functional_allele_calls.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene type: TRGV\n",
      "Number of samples: 47\n",
      "Number of unique functional alleles: 9\n",
      "\n",
      "Loading annotations for TRDV from ../../digger-functional-annotations-all/\n",
      "Gene type: TRDV\n",
      "Number of samples: 33\n",
      "Number of unique functional alleles: 2\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:benchmark_utils:No genotype calls file found for HG02109. Pattern: ../immunotyper-output/20x/trdv/HG02109.final*TRDV_functional_allele_calls.txt\n",
      "ERROR:benchmark_utils:No genotype calls file found for NA21309. Pattern: ../immunotyper-output/20x/trdv/NA21309.final*TRDV_functional_allele_calls.txt\n"
     ]
    }
   ],
   "source": [
    "gene_types = ['IGHV', 'IGLV', 'IGKV', 'TRAV', 'TRBV', 'TRGV', 'TRDV']\n",
    "\n",
    "first = True\n",
    "\n",
    "for gene_type in gene_types:\n",
    "    functional_ground_truths = load_assembly_annotations(gene_type, path=\"../../digger-functional-annotations-all/\")\n",
    "    print(f\"Gene type: {gene_type}\")\n",
    "    print(f\"Number of samples: {len(functional_ground_truths)}\")\n",
    "    print(f\"Number of unique functional alleles: {len(set([allele for alleles in functional_ground_truths.values() for allele in alleles]))}\")\n",
    "    print()\n",
    "\n",
    "    # Create empty lists to store results\n",
    "    results = []\n",
    "\n",
    "    for sample, functional_ground_truth_set in functional_ground_truths.items():\n",
    "\n",
    "        try:\n",
    "            genotypes = load_genotype_calls(sample, gene_type, call_base_dir=call_base_dir)\n",
    "        except FileNotFoundError:\n",
    "            # log.warning(f\"Genotype calls not found for {sample}\")\n",
    "            continue\n",
    "\n",
    "        \n",
    "        tp, fp, fn = calculate_accuracy_metrics(list(set(genotypes)), list(set(functional_ground_truth_set)))\n",
    "        precision, recall = calculate_precision_recall(tp, fp, fn)\n",
    "        \n",
    "        # Store results for this sample\n",
    "        results.append({\n",
    "            'sample': sample,\n",
    "            'precision': precision,\n",
    "            'recall': recall,\n",
    "            'true_positives': tp,\n",
    "            'false_positives': fp,\n",
    "            'false_negatives': fn,\n",
    "            'genotypes': list(set(genotypes)),\n",
    "            'ground_truth': list(set(functional_ground_truth_set)),\n",
    "            'gene_type': gene_type\n",
    "        })\n",
    "\n",
    "    gene_results_df = pd.DataFrame(results)\n",
    "    if first:\n",
    "        results_df = gene_results_df\n",
    "        first = False\n",
    "    else:    \n",
    "        results_df = pd.concat([results_df, gene_results_df], ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "96f16f54-7487-432a-b50e-a1a58a6ee689",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results Summary DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_type</th>\n",
       "      <th>precision_mean</th>\n",
       "      <th>recall_mean</th>\n",
       "      <th>precision_median</th>\n",
       "      <th>recall_median</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGHV</td>\n",
       "      <td>0.822</td>\n",
       "      <td>0.800</td>\n",
       "      <td>0.821</td>\n",
       "      <td>0.803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IGLV</td>\n",
       "      <td>0.955</td>\n",
       "      <td>0.899</td>\n",
       "      <td>0.950</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IGKV</td>\n",
       "      <td>0.814</td>\n",
       "      <td>0.720</td>\n",
       "      <td>0.825</td>\n",
       "      <td>0.724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRAV</td>\n",
       "      <td>0.781</td>\n",
       "      <td>0.924</td>\n",
       "      <td>0.777</td>\n",
       "      <td>0.924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRBV</td>\n",
       "      <td>0.873</td>\n",
       "      <td>0.929</td>\n",
       "      <td>0.889</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TRGV</td>\n",
       "      <td>0.922</td>\n",
       "      <td>0.832</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TRDV</td>\n",
       "      <td>0.289</td>\n",
       "      <td>0.983</td>\n",
       "      <td>0.292</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_type  precision_mean  recall_mean  precision_median  recall_median\n",
       "0      IGHV           0.822        0.800             0.821          0.803\n",
       "1      IGLV           0.955        0.899             0.950          0.901\n",
       "2      IGKV           0.814        0.720             0.825          0.724\n",
       "3      TRAV           0.781        0.924             0.777          0.924\n",
       "4      TRBV           0.873        0.929             0.889          0.938\n",
       "5      TRGV           0.922        0.832             1.000          0.857\n",
       "6      TRDV           0.289        0.983             0.292          1.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "LaTeX Table:\n",
      "\\multirow{2}{*}{Gene type}             & Precision       & Recall          & Precision     & Recall       \\\\\n",
      "                      & (mean)          & (mean)          & (median)      & (median)     \\\\\n",
      "\\midrule\n",
      "\\textit{IGHV} & 0.822 & 0.8 & 0.821 & 0.803 \\\\\n",
      "\\textit{IGLV} & 0.955 & 0.899 & 0.95 & 0.901 \\\\\n",
      "\\textit{IGKV} & 0.814 & 0.72 & 0.825 & 0.724 \\\\\n",
      "\\textit{TRAV} & 0.781 & 0.924 & 0.777 & 0.924 \\\\\n",
      "\\textit{TRBV} & 0.873 & 0.929 & 0.889 & 0.938 \\\\\n",
      "\\textit{TRGV} & 0.922 & 0.832 & 1.0 & 0.857 \\\\\n",
      "\\textit{TRDV} & 0.289 & 0.983 & 0.292 & 1.0 \\\\\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Define all gene types\n",
    "gene_types = ['IGHV', 'IGLV', 'IGKV', 'TRAV', 'TRBV', 'TRGV', 'TRDV']\n",
    "\n",
    "# Calculate statistics for each gene type\n",
    "stats = results_df.groupby('gene_type').agg({\n",
    "    'precision': ['mean', 'median'],\n",
    "    'recall': ['mean', 'median']\n",
    "}).round(3)\n",
    "\n",
    "# Prepare data for DataFrame\n",
    "df_data = []\n",
    "latex_table = \"\"\"\\\\multirow{2}{*}{Gene type}             & Precision       & Recall          & Precision     & Recall       \\\\\\\\\n",
    "                      & (mean)          & (mean)          & (median)      & (median)     \\\\\\\\\n",
    "\\\\midrule\"\"\"\n",
    "\n",
    "# Process each gene type\n",
    "for gene_type in gene_types:\n",
    "    try:\n",
    "        # Check if gene type exists in the stats dataframe\n",
    "        if gene_type in stats.index:\n",
    "            # Get values for LaTeX table\n",
    "            precision_mean = stats.loc[gene_type, ('precision', 'mean')]\n",
    "            recall_mean = stats.loc[gene_type, ('recall', 'mean')]\n",
    "            precision_median = stats.loc[gene_type, ('precision', 'median')]\n",
    "            recall_median = stats.loc[gene_type, ('recall', 'median')]\n",
    "            \n",
    "            # Add to LaTeX table\n",
    "            row = f\"\"\"\n",
    "\\\\textit{{{gene_type}}} & {precision_mean} & {recall_mean} & {precision_median} & {recall_median} \\\\\\\\\"\"\"\n",
    "            latex_table += row\n",
    "            \n",
    "            # Add to DataFrame data\n",
    "            df_data.append({\n",
    "                'gene_type': gene_type,\n",
    "                'precision_mean': precision_mean,\n",
    "                'recall_mean': recall_mean,\n",
    "                'precision_median': precision_median,\n",
    "                'recall_median': recall_median\n",
    "            })\n",
    "        else:\n",
    "            # Gene type not in data, add row with N/A values\n",
    "            row = f\"\"\"\n",
    "\\\\textit{{{gene_type}}} & N/A & N/A & N/A & N/A \\\\\\\\\"\"\"\n",
    "            latex_table += row\n",
    "            \n",
    "            # Add to DataFrame data with NaN values\n",
    "            df_data.append({\n",
    "                'gene_type': gene_type,\n",
    "                'precision_mean': np.nan,\n",
    "                'recall_mean': np.nan,\n",
    "                'precision_median': np.nan,\n",
    "                'recall_median': np.nan\n",
    "            })\n",
    "    except Exception as e:\n",
    "        # Handle any exceptions (missing data, etc.)\n",
    "        row = f\"\"\"\n",
    "\\\\textit{{{gene_type}}} & -- & -- & -- & -- \\\\\\\\\"\"\"\n",
    "        latex_table += row\n",
    "        \n",
    "        # Add to DataFrame data with NaN values\n",
    "        df_data.append({\n",
    "            'gene_type': gene_type,\n",
    "            'precision_mean': np.nan,\n",
    "            'recall_mean': np.nan,\n",
    "            'precision_median': np.nan,\n",
    "            'recall_median': np.nan\n",
    "        })\n",
    "        print(f\"Warning: Error processing {gene_type}: {e}\")\n",
    "\n",
    "# Create DataFrame from collected data\n",
    "results_summary_df = pd.DataFrame(df_data)\n",
    "\n",
    "# Display both the DataFrame and LaTeX table\n",
    "print(\"Results Summary DataFrame:\")\n",
    "display(results_summary_df)\n",
    "\n",
    "print(\"\\nLaTeX Table:\")\n",
    "print(latex_table)\n",
    "\n",
    "# Optionally, you can save the DataFrame to CSV\n",
    "# results_summary_df.to_csv('gene_type_statistics.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38be66d6-ea46-4df4-9a10-2b5cdc9ec803",
   "metadata": {},
   "source": [
    "# CNV Sensitive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8ad152d3-0a3c-4aed-ab87-ac13f3b715c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_type = 'IGLV'\n",
    "\n",
    "# Create empty lists to store results\n",
    "results = []\n",
    "\n",
    "for sample_path in glob(f'data/assembly-annotations/{gene_type}-reference-annotations/*'):\n",
    "    sample = os.path.basename(sample_path)\n",
    "    df, functional_ground_truth_set = load_assembly_annotations(sample, gene_type)\n",
    "    try:\n",
    "        genotypes = load_genotype_calls(sample, gene_type)\n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "    \n",
    "    tp, fp, fn = calculate_accuracy_metrics(genotypes, functional_ground_truth_set, cnv=True)\n",
    "    precision, recall = calculate_precision_recall(tp, fp, fn)\n",
    "    \n",
    "    # Store results for this sample\n",
    "    results.append({\n",
    "        'sample': sample,\n",
    "        'precision': precision,\n",
    "        'recall': recall,\n",
    "        'true_positives': tp,\n",
    "        'false_positives': fp,\n",
    "        'false_negatives': fn,\n",
    "        'genotypes': list(set(genotypes)),\n",
    "        'ground_truth': list(set(functional_ground_truth_set)),\n",
    "        'gene_type': gene_type\n",
    "    })\n",
    "\n",
    "# Create DataFrame from results\n",
    "results_df = pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f8a4d04-d2c0-4749-9e64-c08434637797",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_type = 'TRAV'\n",
    "\n",
    "# Create empty lists to store results\n",
    "results = []\n",
    "\n",
    "for sample_path in glob(f'data/assembly-annotations/{gene_type}-reference-annotations/*'):\n",
    "    sample = os.path.basename(sample_path)\n",
    "    df, functional_ground_truth_set = load_assembly_annotations(sample, gene_type)\n",
    "    try:\n",
    "        genotypes = load_genotype_calls(sample, gene_type)\n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "    \n",
    "    tp, fp, fn = calculate_accuracy_metrics(genotypes, functional_ground_truth_set, cnv=True)\n",
    "    precision, recall = calculate_precision_recall(tp, fp, fn)\n",
    "    \n",
    "    # Store results for this sample\n",
    "    results.append({\n",
    "        'sample': sample,\n",
    "        'precision': precision,\n",
    "        'recall': recall,\n",
    "        'true_positives': tp,\n",
    "        'false_positives': fp,\n",
    "        'false_negatives': fn,\n",
    "        'genotypes': list(set(genotypes)),\n",
    "        'ground_truth': list(set(functional_ground_truth_set)),\n",
    "        'gene_type': gene_type\n",
    "    })\n",
    "\n",
    "trav_results_df = pd.DataFrame(results)\n",
    "\n",
    "# Concatenate the two DataFrames\n",
    "results_df = pd.concat([results_df, trav_results_df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1569cea3-0894-4be6-ac75-a57eea02a4eb",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'gene_type'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Calculate statistics for each gene type\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m stats \u001b[38;5;241m=\u001b[39m \u001b[43mresults_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgroupby\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgene_type\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39magg({\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprecision\u001b[39m\u001b[38;5;124m'\u001b[39m: [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmedian\u001b[39m\u001b[38;5;124m'\u001b[39m],\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecall\u001b[39m\u001b[38;5;124m'\u001b[39m: [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmedian\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m      5\u001b[0m })\u001b[38;5;241m.\u001b[39mround(\u001b[38;5;241m3\u001b[39m)\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# Generate LaTeX table\u001b[39;00m\n\u001b[1;32m      8\u001b[0m latex_table \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\"\"\u001b[39m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;124mmultirow\u001b[39m\u001b[38;5;130;01m{{\u001b[39;00m\u001b[38;5;124m2\u001b[39m\u001b[38;5;130;01m}}\u001b[39;00m\u001b[38;5;130;01m{{\u001b[39;00m\u001b[38;5;124m*\u001b[39m\u001b[38;5;130;01m}}\u001b[39;00m\u001b[38;5;130;01m{{\u001b[39;00m\u001b[38;5;124mGene type\u001b[39m\u001b[38;5;130;01m}}\u001b[39;00m\u001b[38;5;124m             & Precision       & Recall          & Precision     & Recall       \u001b[39m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;130;01m\\\\\u001b[39;00m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;124m                      & (mean)          & (mean)          & (median)      & (median)     \u001b[39m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;130;01m\\\\\u001b[39;00m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;124mmidrule\u001b[39m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;124mtextit\u001b[39m\u001b[38;5;130;01m{{\u001b[39;00m\u001b[38;5;124mIGLV\u001b[39m\u001b[38;5;130;01m}}\u001b[39;00m\u001b[38;5;124m & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIGLV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprecision\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m                & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIGLV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecall\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m                & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIGLV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprecision\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmedian\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m                & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIGLV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecall\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmedian\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;130;01m\\\\\u001b[39;00m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;124mtextit\u001b[39m\u001b[38;5;130;01m{{\u001b[39;00m\u001b[38;5;124mTRAV\u001b[39m\u001b[38;5;130;01m}}\u001b[39;00m\u001b[38;5;124m & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTRAV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprecision\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m                & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTRAV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecall\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m                & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTRAV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprecision\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmedian\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m                & \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstats\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTRAV\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrecall\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmedian\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;124m\"\"\"\u001b[39m\n",
      "File \u001b[0;32m/data/fordmk/miniconda3/lib/python3.9/site-packages/pandas/core/frame.py:9183\u001b[0m, in \u001b[0;36mDataFrame.groupby\u001b[0;34m(self, by, axis, level, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[1;32m   9180\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m level \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m by \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   9181\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou have to supply one of \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mby\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m and \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlevel\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 9183\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameGroupBy\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   9184\u001b[0m \u001b[43m    \u001b[49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9185\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeys\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mby\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9186\u001b[0m \u001b[43m    \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9187\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9188\u001b[0m \u001b[43m    \u001b[49m\u001b[43mas_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mas_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9189\u001b[0m \u001b[43m    \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9190\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroup_keys\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroup_keys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9191\u001b[0m \u001b[43m    \u001b[49m\u001b[43mobserved\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mobserved\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9192\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdropna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdropna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9193\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/data/fordmk/miniconda3/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:1329\u001b[0m, in \u001b[0;36mGroupBy.__init__\u001b[0;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[1;32m   1326\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdropna \u001b[38;5;241m=\u001b[39m dropna\n\u001b[1;32m   1328\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m grouper \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1329\u001b[0m     grouper, exclusions, obj \u001b[38;5;241m=\u001b[39m \u001b[43mget_grouper\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1330\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1331\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1332\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1333\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1334\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1335\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobserved\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mobserved\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mno_default\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mobserved\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1336\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdropna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdropna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1337\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1339\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m observed \u001b[38;5;129;01mis\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mno_default:\n\u001b[1;32m   1340\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28many\u001b[39m(ping\u001b[38;5;241m.\u001b[39m_passed_categorical \u001b[38;5;28;01mfor\u001b[39;00m ping \u001b[38;5;129;01min\u001b[39;00m grouper\u001b[38;5;241m.\u001b[39mgroupings):\n",
      "File \u001b[0;32m/data/fordmk/miniconda3/lib/python3.9/site-packages/pandas/core/groupby/grouper.py:1043\u001b[0m, in \u001b[0;36mget_grouper\u001b[0;34m(obj, key, axis, level, sort, observed, validate, dropna)\u001b[0m\n\u001b[1;32m   1041\u001b[0m         in_axis, level, gpr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m, gpr, \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1042\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1043\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(gpr)\n\u001b[1;32m   1044\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(gpr, Grouper) \u001b[38;5;129;01mand\u001b[39;00m gpr\u001b[38;5;241m.\u001b[39mkey \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1045\u001b[0m     \u001b[38;5;66;03m# Add key to exclusions\u001b[39;00m\n\u001b[1;32m   1046\u001b[0m     exclusions\u001b[38;5;241m.\u001b[39madd(gpr\u001b[38;5;241m.\u001b[39mkey)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'gene_type'"
     ]
    }
   ],
   "source": [
    "# Calculate statistics for each gene type\n",
    "stats = results_df.groupby('gene_type').agg({\n",
    "    'precision': ['mean', 'median'],\n",
    "    'recall': ['mean', 'median']\n",
    "}).round(3)\n",
    "\n",
    "# Generate LaTeX table\n",
    "latex_table = f\"\"\"\\\\multirow{{2}}{{*}}{{Gene type}}             & Precision       & Recall          & Precision     & Recall       \\\\\\\\\n",
    "                      & (mean)          & (mean)          & (median)      & (median)     \\\\\\\\\n",
    "\\\\midrule\n",
    "\\\\textit{{IGLV}} & {stats.loc['IGLV', ('precision', 'mean')]}                & {stats.loc['IGLV', ('recall', 'mean')]}                & {stats.loc['IGLV', ('precision', 'median')]}                & {stats.loc['IGLV', ('recall', 'median')]} \\\\\\\\\n",
    "\\\\textit{{TRAV}} & {stats.loc['TRAV', ('precision', 'mean')]}                & {stats.loc['TRAV', ('recall', 'mean')]}                & {stats.loc['TRAV', ('precision', 'median')]}                & {stats.loc['TRAV', ('recall', 'median')]}\\\\\\\\\"\"\"\n",
    "\n",
    "print(latex_table)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
